socket.go

```diff
@@ -96,16 +99,23 @@ func listenPacket(network, addr string) (pc net.PacketConn, err error) {
 	return net.ListenPacket(network, addr)
 }
 
-func NewSocket(network, addr string) (*Socket, error) {
+func NewSocket(network, addr string, logger log.Logger) (*Socket, error) {
```

Comment:
Add a new constructor that takes a Logger, or make it optional using the optional function argument pattern? There's a lot of code that uses this and it will all break like this.

Reply:
Haven't worked with optional function arguments before in Go, so I'm not sure if this what you meant but: I wrapped `NewSocket` to pass `Logger`, and renamed the previous `NewSocket`, which accepts a mandatory `log.Logger` argument, to `NewSocketWithLogger`.

I can use a different method for optional arguments if you'd like – I think the other main paradigm I've seen is variadic function arguments? – but this should maintain compatibility with previous implementations while allowing folks who care to use custom loggers.

Reply:
Yeah I meant variadic function arguments. Do you want to do that?

Reply:
Just based on experience in other public APIs, when there's a proliferation of options, having a constructor for every combination quickly gets out of control 😆 .

Reply:
Variadic it is!